### DISCUSSION TITLE
Replace `NoCountriesExcept` and `AllCountriesExcept` with country metadata?

### DISCUSSION THREAD

**Discussion by FloEdelmann** [streetcomplete/StreetComplete#4096] [Ideas]
Some quests are enabled/disabled only in certain countries using `val enabledInCountries = NoCountriesExcept(…)` or `val enabledInCountries = AllCountriesExcept(…)` because of a certain feature that is present/missing in these countries.

Would it make sense to extract these metadata to https://github.com/streetcomplete/countrymetadata and then refactor the `enabledInCountries` like this?

#### Currently

```kt
// https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/OsmElementQuestType.kt#L37
val enabledInCountries: Countries get() = AllCountries // always enabled by default

// https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/tactile_paving/TactilePavingUtil.kt#L6
val COUNTRIES_WHERE_TACTILE_PAVING_IS_COMMON = NoCountriesExcept(…)

// https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/tactile_paving/AddTactilePavingCrosswalk.kt#L39
override val enabledInCountries = COUNTRIES_WHERE_TACTILE_PAVING_IS_COMMON

// https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/quests/address/AddHousenumber.kt#L29-L37
override val enabledInCountries = AllCountriesExcept(…)
```

#### My suggestion

```kt
// …/data/osm/osmquests/OsmElementQuestType.kt
fun isEnabled(countryInfo: CountryInfo) = true // always enabled by default

// …/quests/tactile_paving/AddTactilePavingCrosswalk.kt
override fun isEnabled(countryInfo: CountryInfo) = countryInfo.isTactilePavingKnown

// …/quests/address/AddHousenumber.kt
override fun isEnabled(countryInfo: CountryInfo) =
    !countryInfo.housenumbersImported && !countryInfo.housenumbersOutsideBuilding
```

### Benefits

* metadata usable for other applications, too
* one single way of specifying country metadata, regardless whether they are used to enable/disable quests or show country-specific options inside quests
* make sharing filters between multiple quests easier, e.g. for #4095, the existing filter from `AddSummitRegister` wouldn't need to be duplicated or manually extracted into a shared variable (because it is already extracted)
* makes it possible to share filter between quest enabling/disabling and (other) quest UI:
  https://github.com/streetcomplete/StreetComplete/blob/c6ff6bb6c25d55a7d5afddb53b90e7fb87192e11/app/src/main/java/de/westnordost/streetcomplete/quests/recycling_glass/DetermineRecyclingGlass.kt#L26-L27
* makes it possible to enable/disable a quest in a country subdivision (not sure if this is currently supported already)

### Drawbacks

* More effort to enable/disable a quest in certain countries:
  - Create new file in https://github.com/streetcomplete/countrymetadata
  - Add property in https://github.com/streetcomplete/StreetComplete/blob/master/app/src/main/java/de/westnordost/streetcomplete/data/meta/CountryInfo.kt
  - Update metadata (already automated via a Gradle task)
* Enabling a quest only in certain countries because it is not implemented for other countries is not really a metadata worth extracting. Maybe that can be solved by something like:
  ```kt
  // only implemented in Germany for now
  override fun isEnabled(countryInfo: CountryInfo) = countryInfos.countryCodes.contains("DE")
  ```
  Still, it's more verbose.

**Comment by matkoniecz** [comment]
Not entirely sure is it worth it, but in such case metadata should not be `enableTactilePavingOnCrossingQuest` but something more generic like `TactilePavingOnCrossingsIsPopular` (appears in example already - is it possible to do with all quests?)

> metadata usable for other applications, too

is there any user of existing metadata or any real potential for user of new metadata?

**Reply by FloEdelmann** [reply]
https://github.com/streetcomplete/countrymetadata/blob/2cdf73ccbc6168f14c383ed80feb90cea4695c66/README.md?plain=1#L18:

> [A/B Street](https://github.com/a-b-street/abstreet) uses some of this data to localize the rendering of streets

They might maybe be interested in e.g. `isTactilePavingKnown`. But I don't know any other specific use cases.

**Comment by mnalis** [comment]
As a newbie when it comes to Android SC quest writing, I don't really like it.

In essence, it would be making writing **all simpler quests 25% more complex** (need editing 4 files instead of current 3), **in order to make rare already complex quests with shared metadata _slightly_ easier**. 

I do agree that manual copy/paste is wrong way to "share" such filters, but perhaps it is possible to find a solution for sharing filters between such more complex quests, which will need extra work **only** when such more advanced feature are actually required (and not for **all** quests as proposed here)?

Which leaves making metadata usable for other projects. Although it is admirable goal, unless we already have a few of them waiting to jump on that bandwagon right know and urging us to hurry, I'd say it is **not** worth the degradation of simplicity (e.g. like way proposed in https://github.com/streetcomplete/StreetComplete/discussions/3302 for sharing SC code also wasn't worth degradation of simplicity, IMHO)

**Comment by FloEdelmann** [comment]
I've now created draft PRs to better see how it could work and I actually quite like it:

* https://github.com/streetcomplete/StreetComplete/pull/4101
* https://github.com/streetcomplete/countrymetadata/pull/6

@mnalis could you please have a look at them and see whether your concerns are confirmed or dispelled?

**Comment by FloEdelmann** [comment]
_Originally posted by @westnordost in https://github.com/streetcomplete/StreetComplete/issues/4101#issuecomment-1153191975:_

Hmm, I don't really like this.

Country metadata is so far composed of (afair/mostly) hard factual information, such as the start of the workday, the appearance of a cycle lane etc. 
Now, a quest being disabled in StreetComplete is not necessarily based on such facts but a matter of consideration:
- does the given property (cycle track, vibrating button, ...) appear in such numbers that it is worthwhile to ask surveyors about it (💤 no spam)?
- do the mapping practices or (local) tagging guidelines influence the quest in a way that it can't fulfill all the quest guidelines? (e.g. housenumber quest: housenumbers tagged not on buildings; all housenumbers imported for third party etc..)

Trying to shoehorn these reasons into the countrymetadata does not appear so wise because it is both less flexible and means more effort for contributor of a quest.

Additionally, in regards to adding such fuzzy stuff as `usuallyHasXYZ` to countrymetadata: From the point of view of maintenance of this repository, this sounds like a precedent for a nightmare to maintain. IIRC the readme for countrymetadata requests to always give the source (in a comment behind that line) for each contribution made, because that data should be based on hard facts. If there is data that is so subjective that it is a matter of consideration... well, that's to things that are difficult to marry.

**Reply by FloEdelmann** [reply]
Good point, I didn't consider this. Most of the metadata are indeed subjective, but some are verifiable facts:

* `addressHasStreet`
* `busStopIdentifiers`
* `postBoxesHaveCollectionTimes`
* `postBoxesHaveRef`
* `postBoxesHaveRoyalCypher`

Would it still make sense to refactor the way quests are enabled/disabled to allow (but not force) relying on country metadata? That would give the same benefits as listed in the original post, minus reusability in other apps of course.

Also note that the subjective `isUsuallyAnyGlassRecyclableInContainers` already exists and would benefit from that refactoring.

I.e. should I …

1. … keep the `isEnabled(countryInfo: CountryInfo) = …` refactoring, and
2. … write helper functions for `CountryInfo` that mimic the `NoCountriesExcept` and `AllCountriesExcept` logic, depending on the country codes, and
3. … re-inline the subjective quests and refactor them to use the new functions?

Or should I completely abandon the idea and keep everything as is?

**Reply by westnordost** [reply]
I agree, some data is based on verifiable facts. So from this point of view, nothing speaks against including those into `countrymetadata` and a PR that does that won't be rejected. However, I do not like the idea to couple quest disablements with country metadata, so from the implementation in streetcomplete side, I'd opt for to completely abandon the idea.
